### PR TITLE
Allow mutable graphs

### LINKED ISSUE #264: Allow mutable graph spaces (support add_vertex! and rem_vertex!)
Hi there, I used ABM with the `step!` function for simulation, but I found that even if some agents were killed in some steps, the graph in the space did not update. I verified this by calculating the vertices of the network using `nv(model.space.graph)`, and it seems that the output is corresponding to the beginning state of the simulation . Does anyone know why this happened? And I am wondering how can I get to know the properties of the network in each step if I cannot make use of `model.space.graph`? Thanks for your time in advance!

### PR DESCRIPTION
**Pull request by Datseris**
Closes #264

I am having trouble because I don't know anything about LightGraphs and perhaps someone should help me here.

How does the order of node integers change in LightGraphs.jl when you add another one or remove one? If I do `rem_vertex!`, can I just do `deleteat!(model.space.s, n)` and then the correspondence between node numbers in the graph, and node numbers in `model.space.s` stays the same? 

And what about adding new nodes? If I add a new node in lightgraphs, is it automatically becoming the `nv(graph)+1` node? From the documentation of `add_vertex!` it looks like it.

### CODE REVIEW

**Comment by Datseris** [comment]
The documentation of LightGraphs.jl says that:

  This operation has to be performed carefully if one keeps external data structures indexed    
  by edges or vertices in the graph, since internally the removal is performed swapping the     
  vertices v and |V|, and removing the last vertex |V| from the graph. After removal the        
  vertices in g will be indexed by 1:|V|-1.

which is too complicated for me to understand right now :D So I'll leave this PR until an expert of light graphs wants to continue it.

**Comment by yanyuan-wang** [comment]
Hi, here's my understanding.

What `rem_vertex!` does is swapping the index of the last node and that of the one to be removed. Suppose there are 10 nodes in the graph, `rem_vertex!(graph, 2)` will delete the second node and then re-index the last node as 2, while keeping the index of the rest 8 nodes the unchanged. So, this "swapping" operation can make sure that the maximum index is equal to the number of nodes in the graph.

And whenever a new node is added, it will automatically be attached with the index of `nv(graph)+1`.

File: src/spaces/graph.jl [hunk 1]
```diff
@@ -127,3 +138,23 @@ function nearby_positions(
...
+`n` node (and all its edges and agents) are deleted.
+"""
+function rem_node!(n::Int, model::ABM{<:AbstractAgent, <: GraphSpace})
+    for a ∈ agents_in_position(n, model); kill_agent!(a, model); end
+    s = model.space.s
+    V = nv(model)
+    s[V], s[n] = s[n], s[V]
+    pop!(s)
+    LightGraphs.rem_vertex!(model.space.graph, n)
+end
```
Comment (Datseris) [inline src/spaces/graph.jl:160]: @yanyuan-wang does this seem correct to you?

**Comment by Datseris** [comment]
I've fixed this and merging when tests pass.

File: src/spaces/graph.jl [hunk 2]
```diff
@@ -127,3 +137,39 @@ function nearby_positions(
...
+**Warning:** LightGraphs.jl (and thus Agents.jl) swaps the index of the last node with
+that of the one to be removed, while every other node remains as is. This means that
+when doing `rem_node!(n, model)` the last node becomes the `n`-th node while the previous
+`n`-th node (and all its edges and agents) are deleted.
+"""
+function rem_node!(model::ABM{<:AbstractAgent, <: GraphSpace}, n::Int)
+    for id ∈ copy(ids_in_position(n, model)); kill_agent!(model[id], model); end
+    V = nv(model)
+    success = LightGraphs.rem_vertex!(model.space.graph, n)
+    n ∉ 1:V && error("Node number exceeds amount of nodes in graph!")
```
Comment (kavir1698) [inline src/spaces/graph.jl:158]: Lines 156 and 158 should be the first two lines of the function.

Isn't `n > V` faster than `n ∉ 1:V`?

File: src/spaces/graph.jl (hunk 1 above)
Comment (yanyuan-wang) [reply src/spaces/graph.jl]: Yes, that looks good!

File: src/spaces/graph.jl (hunk 2 above)
Comment (Datseris) [reply src/spaces/graph.jl:158]: yes, by 1 boolean operation. `n \in range` reduces to:
```
in(x::Integer, r::AbstractUnitRange{<:Integer}) = (first(r) <= x) & (x <= last(r))
```

**Pull request merged**